File: MarsRover/java/MarsRover.java
/**
 * Mars Rover — Java
 *
 * Tunnel-running game. Rover at the bottom, tunnel scrolls down.
 * Steer left/right to avoid walls.
 */

import java.io.IOException;
import java.util.Random;

public class MarsRover {

    static final int WIDTH     = 60;
    static final int HEIGHT    = 25;
    static final int TUNNEL_W  = 14;
    static final int DELAY_MS  = 70;

    private int[] tunnel = new int[HEIGHT];
    private int roverX;
    private int roverRow = HEIGHT - 2;
    private int score = 0;
    private Random rng = new Random();

    public MarsRover() {
        roverX = WIDTH / 2;
        int left = (WIDTH - TUNNEL_W) / 2;
        for (int i = 0; i < HEIGHT; i++) {
            tunnel[i] = left;
            left += rng.nextInt(3) - 1;
            left = Math.max(1, Math.min(WIDTH - TUNNEL_W - 1, left));
        }
    }

    private void scrollTunnel() {
        /* Shift rows down; new row appears at top */
        System.arraycopy(tunnel, 0, tunnel, 1, HEIGHT - 1);
        int drift = rng.nextInt(5) - 2;
        int left = tunnel[1] + drift;
        left = Math.max(1, Math.min(WIDTH - TUNNEL_W - 1, left));
        tunnel[0] = left;
    }

    private void render() {
        StringBuilder sb = new StringBuilder();
        sb.append("\033[2J\033[H");
        sb.append("  MARS ROVER  |  Score: ").append(score).append('\n');
        for (int r = 0; r < HEIGHT; r++) {
            int tl = tunnel[r];
            int tr = tl + TUNNEL_W;
            for (int c = 0; c < WIDTH; c++) {
                if (r == roverRow && c == roverX) sb.append('A');
                else if (c <= tl || c >= tr) sb.append('#');
                else sb.append(' ');
            }
            sb.append('\n');
        }
        sb.append("  A/D keys to steer  |  Q to quit\n");
        System.out.print(sb);
        System.out.flush();
    }

    private boolean checkCollision() {
        int tl = tunnel[roverRow];
        int tr = tl + TUNNEL_W;
        return roverX <= tl || roverX >= tr;
    }

    public void run() throws IOException, InterruptedException {
        // Set terminal to raw mode via stty
        String[] cmd = {"/bin/sh", "-c", "stty -icanon -echo min 0 < /dev/tty"};
        Runtime.getRuntime().exec(cmd).waitFor();

        try {
            boolean alive = true;
            while (alive) {
                // Input
                while (System.in.available() > 0) {
                    int ch = System.in.read();
                    if (ch == 'q' || ch == 'Q') { alive = false; break; }
                    if (ch == 27) { // escape
                        if (System.in.available() > 0 && System.in.read() == '[') {
                            if (System.in.available() > 0) {
                                int k = System.in.read();
                                if (k == 'D' && roverX > 0) roverX--;
                                if (k == 'C' && roverX < WIDTH - 1) roverX++;
                            }
                        }
                    }
                    if ((ch == 'a' || ch == 'A') && roverX > 0) roverX--;
                    if ((ch == 'd' || ch == 'D') && roverX < WIDTH - 1) roverX++;
                }
                if (!alive) break;

                scrollTunnel();
                score++;

                if (checkCollision()) alive = false;

                render();
                Thread.sleep(DELAY_MS);
            }

            System.out.print("\033[2J\033[H\n  GAME OVER!  Final Score: " + score + "\n\n");
        } finally {
            String[] restore = {"/bin/sh", "-c", "stty sane < /dev/tty"};
            Runtime.getRuntime().exec(restore).waitFor();
        }
    }

    public static void main(String[] args) throws Exception {
        new MarsRover().run();
    }
}

File: MarsRover/cpp/mars_rover.cpp
// ================================================================
// Mars Rover — C++
//
// Tunnel-running game. Rover at the bottom, tunnel scrolls down.
// Steer left/right to avoid walls.
// ================================================================

#include <iostream>
#include <vector>
#include <random>
#include <chrono>
#include <thread>
#include <sstream>
#include <termios.h>
#include <unistd.h>
#include <sys/select.h>

constexpr int WIDTH      = 60;
constexpr int HEIGHT     = 25;
constexpr int TUNNEL_W   = 14;
constexpr int DELAY_MS   = 70;
constexpr char ROVER     = 'A';
constexpr char WALL      = '#';

static bool kbhit() {
    timeval tv{0, 0};
    fd_set fds;
    FD_ZERO(&fds);
    FD_SET(0, &fds);
    return select(1, &fds, nullptr, nullptr, &tv) > 0;
}

static int getch_nb() {
    unsigned char c;
    if (read(0, &c, 1) == 1) return c;
    return -1;
}

int main() {
    // Raw terminal
    termios oldt{}, newt{};
    tcgetattr(0, &oldt);
    newt = oldt;
    newt.c_lflag &= ~(ICANON | ECHO);
    tcsetattr(0, TCSANOW, &newt);

    std::mt19937 rng(static_cast<unsigned>(
        std::chrono::steady_clock::now().time_since_epoch().count()));
    std::uniform_int_distribution<int> drift_dist(-1, 1);
    std::uniform_int_distribution<int> drift_dist2(-2, 2);

    // Init tunnel
    std::vector<int> tunnel(HEIGHT);
    int left = (WIDTH - TUNNEL_W) / 2;
    for (int i = 0; i < HEIGHT; ++i) {
        tunnel[i] = left;
        left += drift_dist(rng);
        left = std::max(1, std::min(WIDTH - TUNNEL_W - 1, left));
    }

    int rover_x = WIDTH / 2;
    int rover_row = HEIGHT - 2;
    int score = 0;
    bool alive = true;

    while (alive) {
        // Input
        while (kbhit()) {
            int ch = getch_nb();
            if (ch == 'q' || ch == 'Q') { alive = false; break; }
            if (ch == 27) {
                if (kbhit() && getch_nb() == '[') {
                    int k = getch_nb();
                    if (k == 'D' && rover_x > 0) --rover_x;
                    if (k == 'C' && rover_x < WIDTH - 1) ++rover_x;
                }
            }
            if ((ch == 'a' || ch == 'A') && rover_x > 0) --rover_x;
            if ((ch == 'd' || ch == 'D') && rover_x < WIDTH - 1) ++rover_x;
        }
        if (!alive) break;

        // Scroll
        tunnel.erase(tunnel.begin());
        left = tunnel.back() + drift_dist2(rng);
        left = std::max(1, std::min(WIDTH - TUNNEL_W - 1, left));
        tunnel.push_back(left);
        ++score;

        // Collision
        int tl = tunnel[rover_row];
        int tr = tl + TUNNEL_W;
        if (rover_x <= tl || rover_x >= tr) alive = false;

        // Render
        std::ostringstream out;
        out << "\033[2J\033[H";
        out << "  MARS ROVER  |  Score: " << score << "\n";
        for (int r = 0; r < HEIGHT; ++r) {
            tl = tunnel[r];
            tr = tl + TUNNEL_W;
            for (int c = 0; c < WIDTH; ++c) {
                if (r == rover_row && c == rover_x) out << ROVER;
                else if (c <= tl || c >= tr) out << WALL;
                else out << ' ';
            }
            out << '\n';
        }
        out << "  Arrow keys to steer  |  Q to quit\n";
        std::cout << out.str() << std::flush;

        std::this_thread::sleep_for(std::chrono::milliseconds(DELAY_MS));
    }

    std::cout << "\033[2J\033[H\n  GAME OVER!  Final Score: " << score << "\n\n";
    tcsetattr(0, TCSANOW, &oldt);
    return 0;
}
